### ISSUE TITLE
Prevent circular dependency issues when using type imports (enforce-module-boundarie)

### ISSUE THREAD

**Issue by JeanMGirard** [nrwl/nx#9083]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
It would be a lot simpler if type imports were not calculated into de project graphs. 

```typescript
import type { MailData, MailAdapter } from "utils-emails"
```
_ESLint:  Circular dependency between "xxxxxxxx" and "xxxxxxxx" detected_

## Motivation
Circular dependencies in EsLint rea really annoying and type imports should simply not be considered as dependencies anyway. 
It's also making it difficult to use some common patterns like facades, adapters, etc....

**Comment by fbartho** [comment]
For what it's worth, I (a random internet person) don't agree with this ticket, a type-import is still a circular dependency.

If you allow type-imports as circular dependencies, it can be really spooky for a dev when a minor change starts triggering an error. Keeping this rule firm and clear will help prevent confusion down the line. Adding an exception for type imports will definitely muddy the waters.

**Comment by FrozenPandaz** [comment]
I agree that `type` imports are still dependencies and should create circular dependencies.

If you would like to ignore the lint rule for that line, you can use `// eslint-disable-next-line`.

We don't currently distinguish those imports differently.

**Comment by nhhockeyplayer** [comment]
My eslint model is legit and the IDE imports red errors shut off with eslint options from error to warn

But the hard build fails... is this Nx or Ivy? Is rthere a way to turn off the compiler from stopping the build for these so development can continue?

`>  NX   Could not execute command because the task graph has a circular dependency`

**Comment by g3tr1ght** [comment]
@meeroslav I would disagree with the statement:
> If you would ever try to publish a project to npm, those circular dependencies would be causing issues

Have you published a project whose only circular dependencies in the source code are types? Aren't types extracted into .d.ts modules and all type imports removed from the meaningful bundle so it's free of circular references?
Why does ECMAScript modules specification support circular dependencies?

> Circular dependency is usually a code smell

I do agree that circular dependency is **usually** a code smell. The key word here is **usually**. Not always. What's more important is **code smell is a subjective** (not objective) characteristic.
NX philosophy is **barrel library exports**, which means you can't use precise import specifiers when going cross-libraries. 
However, do you always `import lodash from 'lodash';` or rather `import merge from 'lodash/merge';`? As you are familiar with npm packages, you should as well be familiar with variability in import specifiers usage within each library (`import foo from 'foo';` or `import bar from 'foo/bar';` or `import fiz from 'foo/bar/fiz';`), unless it's packaged as a single module... oh wait! NX makes us use own libraries as if they consist of a single module but they aren't which actually creates circular dependencies.
If instead (breaking NX philosophy) we were to use precise import specifiers (which a sane person would usually do if one doesn't use NX), it eliminates most (if not all!) circular dependencies, and if i have two modules cross-referencing each other, I would split one of them into two (absolutely no concern here). And we would have awesome dev experience, everything now becomes much faster, speed-of-light type check, linting, jest (or whatever it is for unit tests). Fun fact but these tools spend most of their time on module resolution and loading vs doing the actual work. There is absolutely no added value in `import { bar } from '@my-awesome-org/foo';` vs `import { bar } from '@my-awesome-org/foo/bar';` but extra slowness and harm.

> parts of the project should be extracted to a separate library (e.g. domain models).

Why would we do it? Nobody wants to create meaningless sub-libraries (which don't make sense on their own) just to extract some shared code out of multiple logically created libraries. **Simple projects become littered with a hundred libraries some of which consist of only couple modules**.
Say, I have _users_, _comments_, _scheduler_ libraries encapsulating logically relevant code, which could depend on each other. To avoid circular dependencies (_especially in auxiliary modules like tests it's unavoidable, or results in massive code duplication_), I now need to create the following : _users, comments, scheduler, users-comments-shared, users-scheduler-shared, comments-scheduler-shared, users-comments-scheduler-shared, users-comments-shared-types, users-comments-shared-constants, comments-scheduler-shared-types, **how-many-more??**_ This design decision limiting the structure flexibility has a polynomial complexity in terms of scalability** and is harmful in real large-scale development environment. I created a workaround to allow precise imports from libraries but I can't enable `@nrwl/nx/enforce-module-boundaries` with it, it simply reports error.

As code smell is subjective, every official eslint rule has various configurations available. In my opinion, `@nrwl/nx/enforce-module-boundaries` is not mature enough to be useful, it's trying to do too much, why not splitting it into multiple rules?
Transitive circular dependencies (via something like test modules or Storybook stories) we can't work around. I tried to configure NX eslint rule for `"files": ["*.ts", "*.tsx", "*.js", "*.jsx"]` and added an override for `"files": ["*.stories.tsx", "*.test.tsx", "*.test.ts"]`, but this only works if one doesn't have transitive circular dependencies. _In my opinion it is a bug in the rule itself rather than devs making structural design mistakes_. Why overriding every other eslint rule works as expected, but not NX eslint rule?

I presume that "circular dependency" NX team opinion is actually less of an opinion but rather NX design limitation as everything is built around "caching" mechanism, and instead of making the tool providing basic dev tooling and making caching optional, NX decided to **go cache-first at an expense of loosing basic features** (e.g. multiple configurations which Angular CLI supports). I understand that caching is hard. But why would one want to loose so much needed features just because of caching? Most of the **fundamental web development tools support file-based caching**. Yeah, not project-based, but **why would I need to re-check the entire project and all its dependents if I changed one line in a test module**? At the end of the day, most of the tasks can be completely parallelized, I doubt that "caching" is what everyone here started using NX for. It looks good on paper but other than small personal projects face harsh reality.

**Comment by meeroslav** [comment]
I'm genuinely sorry my answer and our linting rule got you so frustrated, @g3tr1ght.

Some of the statements you mentioned are false (e.g. scaling), some are outdated (e.g. change in test triggering dependant projects) and some are simply against the best practices (e.g. deep imports - which might not be bad if you know what you are doing, we do it ourselves in the Nx repo).

Nx and `@nrwl/nx/enforce-module-boundaries` follow a certain philosophy and were battle-tested on applications of various sizes and companies ranging from small startups to large Fortune 100 companies. Nx works for popular open-source projects and large enterprises with 1000s of projects. But that doesn't mean it should work for you as well. No tool is ever perfect, nor it can be ever built to suit everyone's needs.

If you have proposals on improving it, we welcome you to submit a PR. Otherwise, you might have a specific use case for which our tools simply do not work or work against.

**Comment by g3tr1ght** [comment]
Thanks for getting back to me, @meeroslav.

No offence, but I doubt there does exist an org with 1000s of projects based on multiple frameworks in a single NX monorepo with no workarounds, unless bulk of these "projects" are (like I mentioned above) _users-comments-shared, users-scheduler-shared, comments-scheduler-shared, users-comments-scheduler-shared, users-comments-shared-types, users-comments-shared-constants, comments-scheduler-shared-types_. You won't be able to hit even 100 real projects before 100 separate release schedules will bite you quite hard with unwanted changes popping here and there, merge conflicts and impossible major dependencies migrations (assuming you try to share as much code as possible).

And I do agree that it's impossible to come up with a tool that fits everyone's needs but concerns that I mentioned are really basic scenarios and large Fortune 100 companies do have legacy migrations quite often, not every project is new development.
